src/java/org/apache/cassandra/service/MigrationManager.java

```diff
@@ -77,121 +69,6 @@ public void unregister(MigrationListener listener)
-            maybeScheduleSchemaPull(UUID.fromString(value.value), endpoint, state.getApplicationState(ApplicationState.RELEASE_VERSION).value, onChange);
-    }
-
-    /**
-     * If versions differ this node sends request with local migration list to the endpoint
-     * and expecting to receive a list of migrations to apply locally.
-     */
-    private static void maybeScheduleSchemaPull(final UUID theirVersion, final InetAddress endpoint, String  releaseVersion, boolean onChange)
-    {
-        String ourMajorVersion = FBUtilities.getReleaseVersionMajor();
-        if (!releaseVersion.startsWith(ourMajorVersion))
-        {
-            logger.debug("Not pulling schema because release version in Gossip is not major version {}, it is {}", ourMajorVersion, releaseVersion);
-            return;
-        }
-
-        if ((Schema.instance.getVersion() != null && Schema.instance.getVersion().equals(theirVersion)) || !shouldPullSchemaFrom(endpoint))
-        {
-            logger.debug("Not pulling schema because versions match or shouldPullSchemaFrom returned false");
-            return;
-        }
-
-        if (Schema.emptyVersion.equals(Schema.instance.getVersion()) || runtimeMXBean.getUptime() < MIGRATION_DELAY_IN_MS)
-        {
-            // If we think we may be bootstrapping or have recently started, submit MigrationTask immediately
-            logger.debug("Submitting migration task for {}", endpoint);
-            submitMigrationTask(endpoint);
-        }
-
-        if (onChange && Boolean.getBoolean("palantir_cassandra.unsafe_schema_migration"))
-        {
-            // if we're scheduling due to a schema change (and not because a node is newly alive or joined)
-            // and we're in unsafe_schema_migration mode, then don't submit a MigrationTask
-            logger.warn("Not pulling schema because of schema change while in unsafe schema migration mode");
-            return;
-        }
-        else
-        {
-            // Include a delay to make sure we have a chance to apply any changes being
-            // pushed out simultaneously. See CASSANDRA-5025
-            Runnable runnable = new Runnable()
-            {
-                public void run()
-                {
-                    // grab the latest version of the schema since it may have changed again since the initial scheduling
-                    EndpointState epState = Gossiper.instance.getEndpointStateForEndpoint(endpoint);
-                    if (epState == null)
-                    {
-                        logger.debug("epState vanished for {}, not submitting migration task", endpoint);
-                        return;
-                    }
-                    VersionedValue value = epState.getApplicationState(ApplicationState.SCHEMA);
-                    UUID currentVersion = UUID.fromString(value.value);
-
-                    if (!currentVersion.equals(theirVersion))
```

Comment:
I _think_ this check, which was introduced in this PR: https://github.com/palantir/cassandra/pull/95 is still relevant and should still be included after the refactor if we wanna keep the optimization it provided. Even though `MigrationCoordinator` functions are synchronized and the class only has one instance, you could still receive Gossip schema changes and make changes to the endpoint state map in parallel. Indeed, the update to the endpoint map happens before the MigrationCoordinator one. Code ref [here](https://github.com/palantir/cassandra/blob/164482859c6e6b370ad6c14974612c51142f9e3b/src/java/org/apache/cassandra/service/StorageService.java#L2134)

Reply:
I agree that this is probably a behavior we still want to have, but since the whole codepath is refactored, the check should probably be implemented differently. Optimization is: 

- We've scheduled a schema pull in 60s for a particular remote schema version
- During the wait, the schema has changed, and we schedule another schema pull for a different remote schema version
- We now have two schema pull requests for the same endpoint, so we should only execute one.